src/vcpkg/binarycaching.cpp

```diff
@@ -2417,92 +2420,165 @@ namespace vcpkg
+        if (m_push_thread.joinable())
+        {
+            m_push_thread.join();
+        }
+    }
+
+    void BinaryCache::BinaryCacheData::push_thread_main(ReadOnlyBinaryCacheData* ro_data)
+    {
+        std::vector<ActionToPush> my_tasks;
+        int count_pushed = 0;
+        while (true)
+        {
+            m_actions_to_push.wait_for_items(my_tasks);
+            if (my_tasks.empty())
+            {
+                break;
+            }
+            for (auto& action_to_push : my_tasks)
+            {
+                ElapsedTimer timer;
                 if (m_needs_zip_file)
                 {
-                    Path zip_path = request.package_dir + ".zip";
-                    auto compress_result = m_zip_tool.value_or_exit(VCPKG_LINE_INFO)
-                                               .compress_directory_to_zip(m_fs, request.package_dir, zip_path);
+                    Path zip_path = action_to_push.request.package_dir + ".zip";
+                    auto compress_result =
+                        m_zip_tool.value_or_exit(VCPKG_LINE_INFO)
+                            .compress_directory_to_zip(m_fs, action_to_push.request.package_dir, zip_path);
                     if (compress_result)
                     {
-                        request.zip_path = std::move(zip_path);
+                        action_to_push.request.zip_path = std::move(zip_path);
                     }
                     else
                     {
-                        out_sink.println(Color::warning,
-                                         msg::format_warning(msgCompressFolderFailed, msg::path = request.package_dir)
-                                             .append_raw(' ')
-                                             .append_raw(compress_result.error()));
+                        auto this_line = DiagnosticLine{DiagKind::Warning,
+                                                        msg::format(msgCompressFolderFailed,
+                                                                    msg::path = action_to_push.request.package_dir)}
+                                             .to_message_line();
+                        this_line.print(" ");
+                        this_line.print(compress_result.error());
+                        m_bg_msg_sink.println(std::move(this_line));
                     }
                 }
 
                 size_t num_destinations = 0;
-                for (auto&& provider : m_config.write)
+                for (auto&& provider : ro_data->m_config.write)
                 {
-                    if (!provider->needs_zip_file() || request.zip_path.has_value())
+                    if (!provider->needs_zip_file() || action_to_push.request.zip_path.has_value())
                     {
-                        num_destinations += provider->push_success(request, out_sink);
+                        num_destinations += provider->push_success(action_to_push.request, stdout_sink);
```

Comment:
```suggestion
                        num_destinations += provider->push_success(action_to_push.request, m_bg_msg_sink);
```

Just noting that I noticed this, I have already fixed it.